compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -1889,16 +1892,49 @@ static void buildDefaultReadWriteFunctions(AggregateType* ct) {
   if (isArrayImplType(ct))
     return;
 
-  if (functionExists("writeThis", dtMethodToken, ct, dtAny)) {
+  if (functionExists("deserialize", dtMethodToken, ct, dtAny, dtAny)) {
+    hasDeserialize = true;
+  }
+
+  if (functionExists("serialize", dtMethodToken, ct, dtAny, dtAny)) {
+    hasSerialize = true;
+  }
+
+  // TODO: should these say something about the type declared on?
+  if (FnSymbol* fn = functionExists("writeThis", dtMethodToken, ct, dtAny)) {
+    if (hasSerialize == false) {
+      USR_WARN(fn, "'writeThis' is deprecated. Please use 'serialize' methods instead.");
+    }
     hasWriteThis = true;
   }
 
-  if (functionExists("readThis", dtMethodToken, ct, dtAny)) {
+  if (FnSymbol* fn = functionExists("readThis", dtMethodToken, ct, dtAny)) {
+    if (hasDeserialize == false) {
+      USR_WARN(fn, "'readThis' is deprecated. Please use 'deserialize' methods instead.");
+    }
     hasReadThis = true;
   }
 
-  if (functionExists("serialize", dtMethodToken, ct, dtAny, dtAny)) {
-    hasSerialize = true;
+  forv_Vec(FnSymbol, method, ct->methods) {
+    if (method != nullptr &&
+        method->isInitializer() &&
+        method->numFormals() == 4 &&
+        strcmp(method->getFormal(3)->name, "reader") == 0 &&
+        strcmp(method->getFormal(4)->name, "deserializer") == 0) {
+      readerInit = method;
+      break;
+    }
+  }
+
+  if (hasSerialize || hasDeserialize ||
+      (readerInit != nullptr &&
+       readerInit->hasFlag(FLAG_COMPILER_GENERATED) == false &&
+       ct->getModule()->modTag != MOD_INTERNAL)) {
+    // If there's a user-defined 'serialize' method...
+    // Or a user-defined 'deserialize' method...
+    // Or a user-defined 'init'-deserializing method...
+    // Then do not generate anything (except for hinting compiler errors)
```

Comment:
Is the`deserializeFrom` type method part of the stable interface, and should it be included in this list?

Reply:
``deserializeFrom`` is not stable at this time, and is primarily a workaround for certain types not supporting initializers.